File: source/core/hiddenapi/src/main/java/android/content/pm/PackageManagerHidden.java
package android.content.pm;

import android.os.UserHandle;

import java.util.List;

import dev.rikka.tools.refine.RefineAs;

/**
 * @see <a href="https://cs.android.com/android/platform/superproject/+/android-8.0.0_r51:frameworks/base/core/java/android/content/pm/PackageManager.java">PackageManager.java</a>
 */
@RefineAs(PackageManager.class)
public class PackageManagerHidden {
    public List<PackageInfo> getInstalledPackagesAsUser(int flags, int userId) {
        throw new RuntimeException("Stub!");
    }

    public PackageInfo getPackageInfoAsUser(String packageName, int flags, int userId) {
        throw new RuntimeException("Stub!");
    }

    public void grantRuntimePermission(String packageName, String permissionName, UserHandle user) {
        throw new RuntimeException("Stub!");
    }

    public void revokeRuntimePermission(String packageName, String permissionName, UserHandle user) {
        throw new RuntimeException("Stub!");
    }

    public int getPermissionFlags(String permissionName, String packageName, UserHandle user) {
        throw new RuntimeException("Stub!");
    }

    public void updatePermissionFlags(String permissionName, String packageName, int flagMask, int flagValues, UserHandle user) {
        throw new RuntimeException("Stub!");
    }
}

File: source/core/hiddenapi/src/main/java/android/content/pm/PackageManagerHidden.kt
package android.content.pm

import android.annotation.SuppressLint
import android.os.PersistableBundle
import android.os.UserHandle
import org.lsposed.hiddenapibypass.HiddenApiBypass

@SuppressLint("PrivateApi")
@Suppress("UNCHECKED_CAST")
object PackageManagerHidden {
    fun getInstalledPackagesAsUser(packageManager: PackageManager, flags: Int, userId: Int): List<PackageInfo> =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "getInstalledPackagesAsUser", flags, userId) as List<PackageInfo>

    fun getPackageInfoAsUser(packageManager: PackageManager, packageName: String, flags: Int, userId: Int): PackageInfo =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "getPackageInfoAsUser", packageName, flags, userId) as PackageInfo

    fun grantRuntimePermission(packageManager: PackageManager, packageName: String, permName: String, user: UserHandle) =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "grantRuntimePermission", packageName, permName, user)

    fun revokeRuntimePermission(packageManager: PackageManager, packageName: String, permName: String, user: UserHandle) =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "revokeRuntimePermission", packageName, permName, user)

    fun getPermissionFlags(packageManager: PackageManager, packageName: String, permName: String, user: UserHandle): Int =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "getPermissionFlags", permName, packageName, user) as Int

    fun updatePermissionFlags(packageManager: PackageManager, packageName: String, permName: String, user: UserHandle, flagMask: Int, flagValues: Int) =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "updatePermissionFlags", permName, packageName, flagMask, flagValues, user)

    fun setPackagesSuspended(packageManager: PackageManager, packageNames: Array<String>, suspended: Boolean, appExtras: PersistableBundle?, launcherExtras: PersistableBundle?, dialogInfo: Any?) =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "setPackagesSuspended", packageNames, suspended, appExtras, launcherExtras, dialogInfo)

    fun isPackageSuspended(packageManager: PackageManager, packageName: String): Boolean =
        HiddenApiBypass.invoke(Class.forName("android.content.pm.PackageManager"), packageManager, "isPackageSuspended", packageName) as Boolean
}
